### DISCUSSION TITLE
template folder not installed with plugin

### DISCUSSION THREAD

**Discussion by 0Markus0** [netbox-community/netbox#7414] [Q&A] [answered]
Hello,
I develop a plugin for netbox. My problem is that the template folder is not installed if I install the plugin.
Can someone help me?

**Comment by candlerb** [comment]
I recommend you start by installing the "netbox-animal-sounds" plugin, exactly as it is, so you can see how it all works.

Otherwise, I suggest you give some more information about your problem, preferably link to the code repo if you can.

What exact step fails: is it `python setup.py develop` ? In that case, you need at least to show the contents of your [setup.py](https://netbox.readthedocs.io/en/stable/plugins/development/#create-setuppy), and your directory structure (i.e. which directory contains your templates).

**Reply by 0Markus0** [reply]
Thanks for your quick reply.

I install the plugin from my gitlab page. My directory structure is same as the "netbox-animal-sounds" structure. I also have the same setup.py as the "netbox-animal-sounds" plugin but with my gitlab-link and project name.

I run the installation with "pip install git+gitlab-link". There are no fails at the installation. Unfortunately in the installed package directory is not a "templates" folder.

As a result I get in netbox the error "template does not exist".

**Reply by candlerb** [reply]
Can you show the *exact* error please?  As I described separately: if the error says that "base.html" does not exist, because you tried to inherit from it, then that's a different problem.

You can easily show whether your package has included the template files or not, using `pip show -f <plugin-name>`.  Remember that it will be installed under the venv package directory, not under `/opt/netbox/netbox/templates`.

If `pip show -f` shows the package installed in your global packages directory, e.g. `/usr/local/lib/python3.8/...`, then you forgot to activate the virtualenv before doing the `pip install`.

If the template file is completely missing from the pip package, then your setup.py is wrong.

If you want any further help than that, you'll have to be more specific, by *showing* the error messages you see, *showing* your setup.py and your directory structure, and *showing* the output of `pip show -f <plugin-name>`

**Reply by 0Markus0** [reply]
Sorry for being too inaccurate. Here are the details:

**This is my structure:**
```
- netbox_littleadmin/
   - migrations/
   - templates/
      - netbox_littleadmin/
        - index.html
        - ...
   - ```__init__.py```
   - admin.py
   - models.py
   - navigation.py
   - template_content.py
   - urls.py
   - views
- setup.py
```


**This is my setup.py:**
```
from setuptools import find_packages, setup

setup(
    name='littleadmin',
    version='0.1',
    description='LittleAdmin NetBox plugin',
    url='https://link_to_gitlab/netbox_littleadmin',
    author='Markus',
    license='Apache 2.0',
    install_requires=[],
    packages=find_packages(),
    include_package_data=True,
)
```
**This is what the package contains:**
```
Name: littleadmin
Version: 0.1
Summary: LittleAdmin NetBox plugin
Home-page: https://link_to_gitlab/netbox_littleadmin
Author: Markus
Author-email: UNKNOWN
License: Apache 2.0
Location: /home/0markus0/.venv/netbox/lib/python3.9/site-packages
Requires:
Required-by:
Files:
  littleadmin-0.1.dist-info/INSTALLER
  littleadmin-0.1.dist-info/METADATA
  littleadmin-0.1.dist-info/RECORD
  littleadmin-0.1.dist-info/REQUESTED
  littleadmin-0.1.dist-info/WHEEL
  littleadmin-0.1.dist-info/direct_url.json
  littleadmin-0.1.dist-info/top_level.txt
  netbox_littleadmin/__init__.py
  netbox_littleadmin/__pycache__/__init__.cpython-39.pyc
  netbox_littleadmin/__pycache__/admin.cpython-39.pyc
  netbox_littleadmin/__pycache__/models.cpython-39.pyc
  netbox_littleadmin/__pycache__/navigation.cpython-39.pyc
  netbox_littleadmin/__pycache__/template_content.cpython-39.pyc
  netbox_littleadmin/__pycache__/urls.cpython-39.pyc
  netbox_littleadmin/__pycache__/views.cpython-39.pyc
  netbox_littleadmin/admin.py
  netbox_littleadmin/migrations/__init__.py
  netbox_littleadmin/migrations/__pycache__/__init__.cpython-39.pyc
  netbox_littleadmin/models.py
  netbox_littleadmin/navigation.py
  netbox_littleadmin/template_content.py
  netbox_littleadmin/urls.py
  netbox_littleadmin/views.py 
```
**And the exact error:**
```
TemplateDoesNotExist at /plugins/netbox_littleadmin/
netbox_littleadmin/index.html
```

**Reply by candlerb** [reply]
Aside: you'll get more readable copy-paste result in Github if you precede and follow each block of quoted text with three backticks (` ``` `) on a line of its own.  This will stop `__init__.py` rendering as "__init__.py", for example.  You can edit your existing post to add them.

You should add `zip_safe=False` to your setup.py, as described [here](https://netbox.readthedocs.io/en/stable/plugins/development/#create-setuppy).

However, I don't think that's the problem, and I don't know why setup.py is including "migrations" but not "templates" in your final package.  Setuptools version? File or directory permissions?  Since it's not a problem with Netbox, you might be able to get help on a Python-specific list.

To get help, I suggest you boil this down to a minimal reproducible test case which doesn't contain any of your proprietary code, and then publish the whole repo on gitlab or github.  Then someone else will be able to run your setup.py and probably tell you very quickly why certain files are being excluded.

EDIT: and since you're installing using `pip install git+<gitlab_url>`, I think you should try doing a local install and/or putting it in github instead - just to prove this isn't a bug in gitlab.

**Comment by candlerb** [comment]
Hmm... I just tried netbox-animal-sounds with netbox v3.0.4, and I get an exception which _might_ be the same problem as you're seeing (although you didn't show the exact error you saw):

```
<class 'django.template.exceptions.TemplateDoesNotExist'>

base.html

Python version: 3.8.10
NetBox version: 3.0.4
```

Here are the files installed by the package:

```
(venv) root@netbox3:/opt/netbox# pip show -f netbox-animal-sounds
Name: netbox-animal-sounds
Version: 0.91
Summary: An example NetBox plugin
Home-page: https://github.com/netbox-community/netbox-animal-sounds
Author: Jeremy Stretch
Author-email: UNKNOWN
License: Apache 2.0
Location: /opt/netbox/venv/lib/python3.8/site-packages
Requires:
Required-by:
Files:
  netbox_animal_sounds-0.91.dist-info/INSTALLER
  netbox_animal_sounds-0.91.dist-info/METADATA
  netbox_animal_sounds-0.91.dist-info/RECORD
  netbox_animal_sounds-0.91.dist-info/REQUESTED
  netbox_animal_sounds-0.91.dist-info/WHEEL
  netbox_animal_sounds-0.91.dist-info/top_level.txt
  netbox_animal_sounds/__init__.py
  netbox_animal_sounds/__pycache__/__init__.cpython-38.pyc
  netbox_animal_sounds/__pycache__/admin.cpython-38.pyc
  netbox_animal_sounds/__pycache__/models.cpython-38.pyc
  netbox_animal_sounds/__pycache__/navigation.cpython-38.pyc
  netbox_animal_sounds/__pycache__/template_content.cpython-38.pyc
  netbox_animal_sounds/__pycache__/urls.cpython-38.pyc
  netbox_animal_sounds/__pycache__/views.cpython-38.pyc
  netbox_animal_sounds/admin.py
  netbox_animal_sounds/api/__init__.py
  netbox_animal_sounds/api/__pycache__/__init__.cpython-38.pyc
  netbox_animal_sounds/api/__pycache__/serializers.cpython-38.pyc
  netbox_animal_sounds/api/__pycache__/urls.cpython-38.pyc
  netbox_animal_sounds/api/__pycache__/views.cpython-38.pyc
  netbox_animal_sounds/api/serializers.py
  netbox_animal_sounds/api/urls.py
  netbox_animal_sounds/api/views.py
  netbox_animal_sounds/migrations/0001_initial.py
  netbox_animal_sounds/migrations/__init__.py
  netbox_animal_sounds/migrations/__pycache__/0001_initial.cpython-38.pyc
  netbox_animal_sounds/migrations/__pycache__/__init__.cpython-38.pyc
  netbox_animal_sounds/models.py
  netbox_animal_sounds/navigation.py
  netbox_animal_sounds/template_content.py
  netbox_animal_sounds/templates/.DS_Store
  netbox_animal_sounds/templates/netbox_animal_sounds/animal.html
  netbox_animal_sounds/templates/netbox_animal_sounds/animal_list.html
  netbox_animal_sounds/templates/netbox_animal_sounds/inc/animal_count.html
  netbox_animal_sounds/urls.py
  netbox_animal_sounds/views.py
(venv) root@netbox3:/opt/netbox#
```

There is no `base.html` installed, but the templates say `{% extends 'base.html' %}`, and that's what's causing the error.  The [documentation](https://netbox.readthedocs.io/en/stable/plugins/development/#extending-the-base-template) instead shows `{% extends 'base/layout.html' %}`
 
I changed `animal.html` and `animal_list.html` to use `base/layout.html`, and now it seems to work OK.

So in this case, the problem was not that the templates weren't installed, but that the templates themselves were wrong.  This may or may not be the same problem that you see.

--------

@jeremystretch: I see that the github repo for [netbox-animal-sounds](https://github.com/netbox-community/netbox-animal-sounds), linked from the [pypi page](https://pypi.org/project/netbox-animal-sounds/), appears to have vanished.

I think this is unfortunate.  It would be very helpful to have a working example plugin, updated to be compatible with Netbox v3.

**Comment by 0Markus0** [comment, marked as answer]
Hello,

I have a fix for my problem.
I think the error was in the `setup.py` at `packages=find_packages()` .  It seems that the method `find_packages()` cannot find the template folder. Can someone explain me why? 

So I added a `MANIFEST.in` in the same directory as the `setup.py` which includes the templates folder and I add `include_package_data = True` in my `setup.py`.
This is the `MANIFEST.in`:

``` 
include netbox_littleadmin/templates/netbox_littleadmin/* 
```

So now when I install the plugin with `pip install git+<github_url>` the template folder is included.

**Reply by candlerb** [reply]
Since find_packages is imported from from setuptools, perhaps you have an old version of setuptools?

Did you make your plugin inside its own virtualenv, as the docs recommend?  In that case, you should be able to upgrade setuptools inside that virtualenv.
